### ISSUE TITLE
Allow stricter typed dispatch with redux-thunk

### ISSUE THREAD

**Issue by Buuhuu** [angular-redux/ng2-redux#211]
Unfortunately I failed using ng2-redux in combination with ThunkAction from redux-thunk using typescript.

The problem is, that dispatch expects an Action but ThunkAction does not extend Action.

**Comment by danielfigueiredo** [comment]
You're right, typescript throws error if you try using the ThunkAction
```typescript
type ThunkAction<R, S, E> = (dispatch: Dispatch<S>, getState: () => S, extraArgument: E) => R;
```
There are ways to fix this issue, not sure which one would be the best. An easy fix that wouldn't need to add the `redux-thunk` dependency is make it either action | function:
```typescript
dispatch = <A extends Action>(action: A | Function): any => {
```
It would also work with any other library that has a custom middleware which is able to understand different types of actions rather than Redux.Action

**Comment by Buuhuu** [comment]
I agree with you. But wouldn't that soften the interface contract to much? 

redux-thunk redeclares the Dispatch<S> function of redux (see: https://github.com/gaearon/redux-thunk/blob/master/index.d.ts#L8) to accept ThunkAction. 

I tried to do so with NgRedux

```
declare module 'ng2-redux' {
    export interface NgRedux<S> {
        dispatch<R, E, A extends Action>(action: ThunkAction<R, S, E> | A): any;
    }
}
```

 but failed with  

> error TS2484: Export declaration conflicts with exported declaration of 'NgRedux'

**Comment by SethDavenport** [comment]
I've used ng2-redux with redux-thunk in the past.However ultimately I felt that the ActionService pattern made it redundant for my needs (see [docs](https://github.com/angular-redux/ng2-redux/blob/master/docs/action-creator-service.md).

Since then I've reverted to simple actions and side-effect management using [redux-observable epics](https://github.com/angular-redux/ng2-redux/blob/master/docs/epics.md).

That said, you may have a different use case than me. Note that `ngRedux.dispatch<T>` had a generic argument type; I was able to dispatch thunks by calling `ngRedux.dispatch<any>(thunkOrAction)`

**Comment by SethDavenport** [comment]
If you're uncomfortable with `<any>` you can probably specify a more specific type at the dispatch call site.

**Comment by Buuhuu** [comment]
@SethDavenport: thanks. I already had a look at the docs and sure, it make sense to go for ActionService.

I thought it will not work to override the generic argument type as it says something extending Action.

**Comment by SethDavenport** [comment]
Hmm - I seem to recall it working but I haven't tried in a while.  The typings should match the official redux typings, currently:

```typescript
  /**
   * Dispatches an action. It is the only way to trigger a state change.
   *
   * The `reducer` function, used to create the store, will be called with the
   * current state tree and the given `action`. Its return value will be
   * considered the **next** state of the tree, and the change listeners will
   * be notified.
   *
   * The base implementation only supports plain object actions. If you want
   * to dispatch a Promise, an Observable, a thunk, or something else, you
   * need to wrap your store creating function into the corresponding
   * middleware. For example, see the documentation for the `redux-thunk`
   * package. Even the middleware will eventually dispatch plain object
   * actions using this method.
   *
   * @param action A plain object representing “what changed”. It is a good
   *   idea to keep actions serializable so you can record and replay user
   *   sessions, or use the time travelling `redux-devtools`. An action must
   *   have a `type` property which may not be `undefined`. It is a good idea
   *   to use string constants for action types.
   *
   * @returns For convenience, the same action object you dispatched.
   *
   * Note that, if you use a custom middleware, it may wrap `dispatch()` to
   * return something else (for example, a Promise you can await).
   */
  dispatch: Dispatch<S>;
```

Where `Dispatch<S>` is defined as 

```typescript
/**
 * A *dispatching function* (or simply *dispatch function*) is a function that
 * accepts an action or an async action; it then may or may not dispatch one
 * or more actions to the store.
 *
 * We must distinguish between dispatching functions in general and the base
 * `dispatch` function provided by the store instance without any middleware.
 *
 * The base dispatch function *always* synchronously sends an action to the
 * store’s reducer, along with the previous state returned by the store, to
 * calculate a new state. It expects actions to be plain objects ready to be
 * consumed by the reducer.
 *
 * Middleware wraps the base dispatch function. It allows the dispatch
 * function to handle async actions in addition to actions. Middleware may
 * transform, delay, ignore, or otherwise interpret actions or async actions
 * before passing them to the next middleware.
 */
export interface Dispatch<S> {
    <A extends Action>(action: A): A;
}
```

In our case, since we're exposing something that is itself a wrapper to the middleware-wrapped native `dispatch`, it may make sense for us to loosen the type restriction a bit.

**Comment by danielfigueiredo** [comment]
Tried using any and it works, but then you loose the ThunkAction type, trying to use ThunkAction type it complains since you can't override the Action type:
```
error TS2344: Type '(dispatch: Dispatch<any>, getState: () => any, extraArgument: any) => any' does not satisfy the constraint 'Action'.
  Property 'type' is missing in type '(dispatch: Dispatch<any>, getState: () => any, extraArgument: any) => any'
```
Maybe making the dispatch generic argument not extends from Action?

**Comment by born2net** [comment]
I am investigating this lib but I currently rely heavily on Thunk ... I am wondering if Thunk is not compatible with this lib?
regards

Sean

**Comment by SethDavenport** [comment]
Redux thunk works fine with ng2-redux - you just have to dispatch like this: `ngRedux.dispatch<any>(thunkOrAction)`.

I should rename this issue - it's more about whether we can do stricter typing than that.

**Comment by born2net** [comment]
Oh I see, tx for the reply, I will test it out shortly.

Angular 2 Kitchen sink: http://ng2.javascriptninja.io
and source@ https://github.com/born2net/ng2Boilerplate
Regards,

Sean

**Comment by born2net** [comment]
I'm wondering if it can be fixed as it's been a while and seems pretty basic

**Comment by SethDavenport** [comment]
This has been a low priority since we've moved away from Thunk for side-effect handling in most of our projects in favour of epics.

**Comment by born2net** [comment]
ok but still can we please get it fixed, as some still use thunks

**Comment by SethDavenport** [comment]
So there are a few reasons thunks are a low priority.  One is that for complex apps we're using Observable epics for side effect handling.

The second, however, is that even before we moved to epics, we realized that _thunks are totally unnecessary with Angular2_. Think about it: the only reason you make a thunk is so you can access `getState` and `dispatch` inside your action creator.

If you use the [ActionCreatorService](https://github.com/angular-redux/ng2-redux/blob/master/docs/action-creator-service.md) pattern instead, you can implement almost the exact same pattern simply by injecting `NgRedux` with Angular2's dependency injector.

Consider a component that uses a thunk for a conditional action creator:

```typescript
import { Injectable } from '@angular/core';
import { NgRedux } from 'ng2-redux';
import { IAppState } from '../store';

@Injectable()
export class CounterActions {
  static INCREMENT_COUNTER: string = 'INCREMENT_COUNTER';

  incrementIfOddUsingThunk() {
    return (getState, dispatch) => {
      const { counter } = getState();

      if (counter % 2 !== 0) {
        return { type: CounterActions.INCREMENT_COUNTER };
      }
    }
  }
}
```

```typescript
import { Component } from '@angular/core';
import { NgRedux, select } from 'ng2-redux';
import { Observable } from 'rxjs/Observable';
import { CounterActions } from '../actions/counter.actions';
import { IAppState } from '../store';

@Component({
  selector: 'counter',
  template: `
  <p>
    Clicked: {{ counter$ | async }} times
    <button (click)="onClickThunkyButton()">Increment if odd using thunk</button>
  </p>
  `
})
export class Counter {
  @select('counter') counter$: Observable<number>;

  constructor(
    public actions: CounterActions,
    private ngRedux: NgRedux<IAppState>) {}

  onClickThunkyButton() {
    this.ngRedux.dispatch(
      this.actions.incrementIfOddUsingThunk());
  }
}
```

Now consider the same action implemented using an Action Creator Service:

```typescript
import { Injectable } from '@angular/core';
import { NgRedux } from 'ng2-redux';
import { IAppState } from '../store';

@Injectable()
export class CounterActions {
  constructor(private ngRedux: NgRedux<IAppState>) {}

  static INCREMENT_COUNTER: string = 'INCREMENT_COUNTER';

  incrementIfOdd() {
    const { counter } = this.ngRedux.getState();

    if (counter % 2 !== 0) {
      this.ngRedux.dispatch({ type: CounterActions.INCREMENT_COUNTER });
    }
  }
}
```

```typescript
import { Component } from '@angular/core';
import { select } from 'ng2-redux';
import { Observable } from 'rxjs/Observable';
import { CounterActions } from '../actions/counter.actions';

@Component({
  selector: 'counter',
  template: `
  <p>
    Clicked: {{ counter$ | async }} times
    <button (click)="actions.incrementIfOdd()">Increment if odd using action creator service</button>
  </p>
  `
})
export class Counter {
  @select('counter') counter$: Observable<number>;

  constructor(public actions: CounterActions) {}
}
```

It's the same logic, and a bit more concise.

**Comment by SethDavenport** [comment]
That said, if you still really _really_ need to continue using thunks, give `ng2-redux@4.2.0-thunkexp.0` a try and let me know.

**Comment by born2net** [comment]
I totally understand. however I am deeply invested in thunks and would need it to transition... plus personally I enjoy working with thunks, it seems very natural to me... tx again and will check thunkexp..

Sean

**Comment by SethDavenport** [comment]
If it fixes your issue I'll roll it into 4.2.0.

**Comment by born2net** [comment]
@SethDavenport good job!
I can confirm no more TS errors with ```ng2-redux@4.2.0-thunkexp.0```

**Comment by born2net** [comment]
now if we just had support for operators inside decorators I would be golden ;)

**Comment by SethDavenport** [comment]
Fixed and published in 4.2.0.

The decorator thing might take a little longer. Need to spec it out carefully.

**Comment by born2net** [comment]
tx so much.
have you had a chance to think about allowing operators inside the selector?
regards
